### PR TITLE
[artifact] only inspect artifact contents

### LINKED ISSUE #26865: Tar artifact without symlink triggers sandbox escape check
### Nomad version
1.10.5

### Operating system and Environment details

Arch Linux

### Issue

When one tries to download a tar artifact with filesystem isolation disabled the deployment fails the artifact inspection introduced in #26608.

### Reproduction steps

Create a tar file and upload it somewhere:

```bash
touch bar
tar --owner=0 --group=0 cvf foo.tar ./bar
```

The created tar file should look like this:
```
./bar0000664000000000000000000000000015067231533010375 0ustar  rootroot
```

Disable the filesystem isolation on the nomad client you test this on:
```hcl
client {
  artifact {
    disable_filesystem_isolation = true
    disable_artifact_inspection = false
  }
}
```

Run the following job file:

```hcl
job "test" {
  type = "service"

  group "test" {
    count = 1

    task "test" {
      driver = "exec"
      config {
        command = "sleep"
        args = [
          "inf"
        ]
      }

      artifact {
        source = "https://fb.hash.works/UWmjs/foo.tar"
      }

      resources {
        cpu   = 100
        memory = 100
      }
    }
  }
}
```

#### Expected Result

The artifact inspection succeeds since the artifact does not link outside the alloc root.

#### Actual Result

The artifact inspection fails:
```
failed to download artifact "https://fb.hash.works/UWmjs/foo.tar": artifact includes symlink that resolves outside of sandbox
```

### Tests

I tried to replicate this as a go test in [`sandbox_test.go`](https://github.com/hashicorp/nomad/blob/main/client/allocrunner/taskrunner/getter/sandbox_test.go#L116), but this is green:
```go
	t.Run("no-symlink-in-tar", func(t *testing.T) {
		artifact := &structs.TaskArtifact{
			RelativeDest: "local/no-symlink",
			GetterSource: "https://fb.hash.works/UWmjs/foo.tar",
		}

		ac := artifactConfig(2 * time.Second)
		sbox := New(ac, logger)

		_, taskDir := SetupDir(t)
		env := noopTaskEnv(taskDir)
		sbox.ac.DisableFilesystemIsolation = true

		err = sbox.Get(env, artifact, "nobody")
		must.NoError(t, err)
	})
```

### PR DESCRIPTION
**Pull request by chrisroberts**
### Description

Artifact inspection would inspect the alloc directory for symlinks
pointing outside of the alloc directory. This can be problematic
for drivers like the raw exec driver which provide the root system
and will likely include symlinks.

Inspecting only the artifact after being fetched can be difficult if
the artifact is fetched to a pre-existing destination as it would
require discerning between new and old paths. Instead, if an artifact
is to be inspected, it is fetched to a temporary location. The isolated
artifact is then inspected before being moved to its final destination.

The behavior of the artifact getter will be inconsistent when artifact
inspection is enabled versus when it is disabled. This is due to the 
behavior difference of go-getter with different sources. Because
go-getter does not behave consistently across different source types,
there will be situations where fetching errors may be encountered
when inspection is disabled but fetching is successful when enabled.

### Testing & Reproduction steps

Reproduction steps are described in #26865

### Links

Fixes #26865

### Contributor Checklist
- [ ] **Changelog Entry** If this PR changes user-facing behavior, please generate and add a
  changelog entry using the `make cl` command.
- [ ] **Testing** Please add tests to cover any new functionality or to demonstrate bug fixes and
  ensure regressions will be caught.
- [ ] **Documentation** If the change impacts user-facing functionality such as the CLI, API, UI,
  and job configuration, please update the  Nomad website documentation to reflect this. Refer to
  the [website README](../website/README.md) for docs guidelines. Please also consider whether the
  change requires notes within the [upgrade guide](../website/content/docs/upgrade/upgrade-specific.mdx).

### Reviewer Checklist
- [ ] **Backport Labels** Please add the correct backport labels as described by the internal
  backporting document.
- [ ] **Commit Type** Ensure the correct merge method is selected which should be "squash and merge"
  in the majority of situations. The main exceptions are long-lived feature branches or merges where
  history should be preserved.
- [ ] **Enterprise PRs** If this is an enterprise only PR, please add any required changelog entry
  within the public repository. 

- [ ] If a change needs to be reverted, we will roll out an update to the code within 7 days.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

### CODE REVIEW

File: client/allocrunner/taskrunner/getter/util.go [hunk 1]
```diff
@@ -189,44 +214,96 @@ func (s *Sandbox) runCmd(env *parameters) error {
...
+	// the artifact contents will have the owner set correctly
+	// but the destination directory will not, so set that now
+	// if it was configured
+	if env.Chown {
+		if err := chownDestination(finalDest, env.User); err != nil {
+			return err
+		}
+	}
+
+	if err := mergeDirectories(env.Destination, finalDest); err != nil {
```
Comment (mismithhisler) [inline client/allocrunner/taskrunner/getter/util.go]: Is the purpose of this merge so that we don't error or override contents that may already exist in the `finalDest`?

Comment (chrisroberts) [reply client/allocrunner/taskrunner/getter/util.go]: So go-getter has two behaviors based on the source used:

1. unpack/place artifact files within an existing directory structure (overwriting files on collision)
2. delete destination directory and unpack/place artifact files

This is using the first approach (and gets into some of the behavior inconsistencies noted in the description).

Comment (tgross) [reply client/allocrunner/taskrunner/getter/util.go]: If I'm understanding the PR description correctly, this is going to change the behavior of those sources that use option 2, isn't it?

The [artifact.destination](https://developer.hashicorp.com/nomad/docs/job-specification/artifact#destination) is placed relative to the task _working directory_. So couldn't we potentially solve this by having the "root" of the check be the task working directory instead of the allocdir?

Comment (chrisroberts) [reply client/allocrunner/taskrunner/getter/util.go]: Yes, that is correct that it will change the behavior of those sources using option 2, but only when artifact inspection is triggered. On platforms where landlock is available and is not disabled, artifact inspection will not be performed.

Isolating the check to the task directory won't be sufficient when using the exec driver, as an example. In that case the task directory will include things from the host (like `bin/`) which will contain symlinks.

Comment (tgross) [reply client/allocrunner/taskrunner/getter/util.go]: > Yes, that is correct that it will change the behavior of those sources using option 2, but only when artifact inspection is triggered. On platforms where landlock is available and is not disabled, artifact inspection will not be performed.

I think that's an ok tradeoff, but we should document that behavior in the upgrade guide, because that's going to break _somebody's_ jobspecs somewhere.

**Review by aimeeu — APPROVED** [review]

**Comment by aimeeu** [comment]
@chrisroberts  Because of this behavior change, do we need to update https://developer.hashicorp.com/nomad/docs/configuration/client#disable_filesystem_isolation to something like:


```
- `disable_filesystem_isolation` `(bool: false)` - Specifies whether to disable
  filesystem isolation for artifact downloads. Applies only to systems where
  filesystem isolation via [landlock] is possible (Linux kernel 5.13+).When
  applied, Nomad merges the unpacked contents of the artifact into the
  destination path if it already exists. Depending on the remote source being
  used, this may be different behavior than what happens  when the artifact is
  unpacked without inspection.
```

File: website/content/docs/upgrade/upgrade-specific.mdx [hunk 2]
```diff
@@ -38,6 +38,15 @@ being silently ignored. Any existing policies with duplicate or invalid keys
...
 to be valid before it can be written to Nomad.
 
+#### Artifact unpacking behavior changes when inspection enabled
+
+Nomad 1.11.0 updates how artifact contents are unpacked and inspected. Artifact
+inspection is applied when `client.disable_filesystem_isolation` is set to `true`
+or the host platform does not support filesystem isolation. When applied, the
+unpacked contents of the artifact will be merged into the destination path if
+it already exists. Depending on the remote source being used, this may be different
+behavior than seen when the artifact is unpacked without inspection.
```
Comment (tgross) [inline website/content/docs/upgrade/upgrade-specific.mdx]: Can we call out the specific remote sources where this is known? We want users to be able to read this and have actionable insights on whether they have to go updating their jobs.

**Comment by aimeeu** [comment]
@chrisroberts Since the docs content has migrated to the web-unified-docs repo, the updates in this PR need to be recreated in the other repo. I can help or do that, assuming this code is going into 1.11.1.

**Comment by chrisroberts** [comment]
Getting back to this: looking at documenting the discrepancies in behavior from using go-getter directly vs redirecting to a temporary directory and then merging in the contents resulted in a large number of differences compounded by edge cases (like a relative destination already existing as a file). Instead, I updated how the temporary destination is setup so that it better mimics the real destination. This allows actual behavior of go-getter to be properly surfaced and stay consistent while still utilizing a temporary directory to allow for inspection. A temporary file within the temporary directory is used to determine if go-getter removed the directory prior to fetch, which allows providing the same behavior.

Added tests are not exhaustive of all the various supported sources and permutations, rather it is demonstrative of the various edge cases to be encountered and shows the same behavior between direct go-getter usage and inspection usage where a temporary directory is first utilized.

File: client/allocrunner/taskrunner/getter/util.go [hunk 3]
```diff
@@ -189,44 +254,118 @@ func (s *Sandbox) runCmd(env *parameters) error {
...
+		}
+
+		// remove the destination and move the source
+		if err := os.RemoveAll(dst); err != nil {
+			return err
+		}
+
+		if err := os.Rename(src, dst); err != nil {
+			return err
+		}
```
Comment (tgross) [inline client/allocrunner/taskrunner/getter/util.go]: Can we atomically update `dst` with `os.Rename` and no removal if it's not a directory?

File: client/allocrunner/taskrunner/getter/util.go [hunk 4]
```diff
@@ -312,3 +451,21 @@ func isPathWithin(rootPath, toCheckPath string) (bool, error) {
...
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if s.IsDir() {
+		return os.RemoveAll(path)
+	}
+
+	return os.Remove(path)
```
Comment (tgross) [inline client/allocrunner/taskrunner/getter/util.go]: If you peek in the stdlib source, `os.RemoveAll` always calls `os.Remove` first, so we can skip this condition because it doesn't get us anything. But if we do that we'll see that `RemoveAll` _also_ returns `nil` if the path doesn't exist, so I think this entire function can be replaced with a call to `os.RemoveAll`.

https://pkg.go.dev/os#RemoveAll
> RemoveAll removes path and any children it contains. It removes everything it can but returns the first error it encounters. If the path does not exist, RemoveAll returns nil (no error). If there is an error, it will be of type [*PathError](https://pkg.go.dev/os#PathError).

File: client/allocrunner/taskrunner/getter/util.go [hunk 5]
```diff
@@ -165,6 +165,71 @@ func (s *Sandbox) runCmd(env *parameters) error {
 	// find the nomad process
 	bin := subproc.Self()
 
+	// If the artifact needs to be inspected, it must first be fetched
+	// to a temporary location for inspection, then moved to the configured
+	// destination.
+	var finalDest string       // value only set if inspection performed
+	var testRecreateDir string // test path to inform if existing destination directory should be removed
+	if !env.DisableArtifactInspection && (env.DisableFilesystemIsolation || !lockdownAvailable()) {
```
Comment (tgross) [inline client/allocrunner/taskrunner/getter/util.go:176]: Given that we're in a case where we know there's no filesystem isolation, is it possible to lean on https://pkg.go.dev/os#OpenRoot for most of these filesystem method calls so that we're at least ensuring we're definitely not chasing symlinks out of the directory?

File: client/allocrunner/taskrunner/getter/sandbox_test.go [hunk 6]
```diff
@@ -119,13 +123,20 @@ func TestSandbox_Get_inspection(t *testing.T) {
 	testutil.RequireRoot(t)
 	logger := testlog.HCLogger(t)
 
-	// Create a temporary directory directly so the repos
-	// don't end up being found improperly
-	tdir, err := os.MkdirTemp("", "nomad-test")
-	must.NoError(t, err, must.Sprint("failed to create top level local repo directory"))
+	sandboxSetup := func() (string, *Sandbox, interfaces.EnvReplacer) {
+		testutil.RequireRoot(t)
```
Comment (tgross) [inline client/allocrunner/taskrunner/getter/sandbox_test.go]: ~If we've disabled filesystem isolation, shouldn't we be able to avoid requiring root?~ Oh we need it for `os.Chown`. Can we leave a comment about that? It'll make it easier to audit tests that need root in the future.
